First prototype of external filter pluggability

azito122 opened the pull request:
In `settings.php`, we:

1. Use a recursive directory iterator to find all files named `fcl_filter.php` and require them
2. Look for any classes matching the pattern `*fcl_external_filter` and get some info about them:
  - Static method `getshortname` returns filter identifier (e.g. `starred`)
  - Static method `getfullname` returns filter full name (e.g. `Starred Courses Filter`)
  - Static method `getcomponent` returns component (e.g. `local_starred_courses`)
  - Use `ReflectionClass` to get the file path in which the class was defined ([see here](https://stackoverflow.com/questions/2420066/finding-the-php-file-at-run-time-where-a-class-was-defined))
3. Use this info to build the multi-select options array
  i. We store shortname and path in the options array key (which will be the value stored in the config table), delimited by pipes (`|`)
  ii. We use labels like: "Filter Name (component_name)"

Then, in `block_filtered_course_list.php`, I abstracted the task of getting the class name for a filter into a lib function (in `lib.php`). This lib function (`get_filter()`) accepts the filter name as an argument (e.g. `shortname` or `starred`) and does the following:

1. Check and return the base filter class if this is a base filter
2. If this is an external filter:
  i. Checks name against the enabled external filters (from settings)
  ii. If found, parse info from config
  iii. Require path and return class name

This method of storing the path within the config value is a bit messy, but, it does mean that there is comparatively little computational load placed within the block building code (as opposed to storing this all in a DB table).

I will write Behat tests once we get this more hashed out.

kwiliarty commented:
I really like the way you're tackling this. 

What would you think about changing the class name expectation just to `*_fcl_filter`? That would match the filename.

What would you think about adding the static functions that we're going to require to the abstract class that we already have in `filter.php`? That would require us to provide them for the native classes as well, but maybe that's okay.

Please add a `$` sign to the end of your preg match string so that we are sure that we are looking at the end of the path. We don't want to pick up things like `fcl_filter.php.txt` etc.

azito122 commented:
Changes made. If you want to test this, local_starred_courses master has an `fcl_filter.php`

azito122 commented:
I was running this on master, but I don't think that should affect it. What problems were you having, exactly?

1. Good point
2. Ok
3. I don't see that there would be a way to do that that would be simpler than just defining it.
4. I think that should be covered by the class exists check (the filter will just return null).

kwiliarty commented:
I've been getting a lot of different errors, but I bet that some of them are just noise. It's probably best if I let you work out the getters for a start. I see this on the admin page:

> Exception - Using $this when not in object context

azito122 commented:
Yes, that would be from the problem you pointed out, which should now be irrelevant. Also there was a thing I missed, which should also now be fixed. So the latest commit should work, theoretically.

kwiliarty commented:
Okay, with latest of everything I now have a working external filter. 👍

kwiliarty commented:
Here's a very small thing: When you call `get_filter` in `block_filtered_course_list.php` you could pass the settings (`$this->fclconfig->externalfilters`) as a second argument so that you don't need to read them from the database again.

kwiliarty commented:
Do you have any thoughts on how we can direct someone to the syntax for an external filter?

kwiliarty commented:
To a couple other things you mentioned above:
>     3. I don't see that there would be a way to do that that would be simpler than just defining it.
Agreed. Might have been nice, but I don't think it's there.
>     4. I think that should be covered by the class exists check (the filter will just return null).
Cool. I was getting a lot of noise for a while, but I agree.

kwiliarty commented:
I'm thinking that the help string under the selector for the external filters might be the place to direct users to the providing module for documentation.

azito122 commented:
How do you mean? Like direct them to the plugin(s) providing external filters?

Also, I think we should use a multicheckbox instead of a multiselect -- it makes more sense for the kind of enable/disable pattern we're imagine.

And finally, how do we want to handle the setting when there are no external filters? The default is, if the options array is empty, not to display the setting at all.

kwiliarty commented:
> How do you mean? Like direct them to the plugin(s) providing external filters?

Yes, even if we just say something like: 'Please refer to the individual external plugins for details on how to use their filters.'

> Also, I think we should use a multicheckbox instead of a multiselect -- it makes more sense for the kind of enable/disable pattern we're imagine.

That makes sense.

> And finally, how do we want to handle the setting when there are no external filters? The default is, if the options array is empty, not to display the setting at all. 

I think that is a good solution.

azito122 commented:
Ok, using multicheckbox now. The lang string I added before directs people to consult the external plugin. Also, I changed the label to be "Filter Name (shortname) from component_plugin".

kwiliarty commented:
That looks great. (I had an obsolete version when I mentioned the help string.)

azito122 commented:
I have no particular objections to implementing an interface as described, but might it be simpler to just return to the getter method pattern that I had originally? That is, have static methods that fetch properties of the class, and then subclasses define the properties rather than the methods. I think that'd be simpler to read than the interface pattern, but I'm open to either.

kwiliarty commented:
I take your point, and I'm open to persuasion. What I had been hoping to achieve was to require the filters to meet certain expectations. When you extend an abstract class, PHP will require you to define any abstract methods declared in the parent class. Failing to do so will generate a PHP error. There's nothing comparable for properties. Of course, in the end, if the required properties are missing or wrong, the code won't work and the developer will probably get an error of a different kind, so maybe that accomplishes nearly the same thing. The interface workaround makes the architecture less clear, and I haven't tried it myself, so I'm not positive it will work. 

The other thing that's on my mind is that we only need the workaround for 5.6 which will not be supported much longer anyway. I could see that being an argument for either perspective. Here's the case for the abstract method approach: If we use properties now it will be hard to switch later because downstream filters would break. If we use an interface now, I believe we can remove it when we stop supporting Moodle 3.3 without a disruption for downstream filters. 

I'm just trying to play out the pros and cons. I still lean slightly toward using abstract methods. Where are you at this point?

azito122 commented:
I'm mostly convinced. Although, I note that this means that a broken external filter will kill your entire site, which I'm now thinking might not be ideal.

kwiliarty commented:
Okay, so why don't you try it out to see if it even works as advertised. This is not committing to it, but just collecting another data point.

In that vein I wonder if you have thought about versions and compatibility. Sooner or later it seems that we're going to need a way for an external filter to report what version of the abstract filter it expects to work with and a way for the main block to gracefully skip over an external filter that is expecting some other version. Might that be another abstract static method?

azito122 commented:
Latest commit includes the interface workaround, and it does indeed work as advertised, which is cool.

Version syncing will be necessary, I agree, but it's a hairy problem to solve. An abstract method like `minimumfclversion()` would sort of work, but also not really, because if we update FCL with a change that makes filters incompatible, a minimum version requirement won't catch that. We could do a max version requirement instead, but external filters can't predict what versions they'll be able to work with.

Ultimately, I feel like FCL has to be the one to determine if a filter is compatible. But I'm not really sure what is the best way to accomplish that.

azito122 commented:
A potential solution would be to maintain a separate version number that represents the "API" version of the external filter interface, in a sense. You could bump it every time you make a change that would invalidate external filters, and external filters would define something like `static function getinterfaceversion()` or whatever. Then we'd do a check like `$classname::getinterfaceversion() === FCL_EXFILTER_INTERFACE_VERSION`

I'm not sure I love that, but it would solve the problem, I think.

kwiliarty commented:
That sounds reasonable to me. If you have other ideas, I'm definitely open. It feels like this is getting quite close now.

azito122 commented:
Alright, how's that look? What pattern do you want to use for the version sync number? Semver?

I've also discovered that if there are no options to display for a multicheckbox, and it is hidden, when you click Save changes, _the config value will not be updated to be blank_. It will maintain whatever value was last entered. This is obviously a problem, since the majority of the checking for the filters is done in settings.php. I'm working on a fix.

azito122 commented:
Done.
My workaround for the problem with the config not being set to blank is to call `set_config` and set it to blank manually if the filter checks on the settings page fail -- how do you feel about that?

kwiliarty commented:
That seems pretty reasonable to me. So here's a question. Let's say I have enabled and external filter and then I remove the plugin that provides it...

azito122 commented:
Then if you visit the setting page it will be gone, and even if you don't, the `file_exists` check in the block code will catch it.

kwiliarty commented:
Excellent.
